csrc/fusion_segmenter.cpp

```diff
@@ -3850,6 +3850,35 @@ void SegmentCandidateFinder::findSegments() {
   }
 }
 
+// Decides whether we should forward an input (or a forwarded input) of a
+// fusion. Currently, we forward an input only when its single use is a UnaryOp.
+// Therefore, this function returns `v`'s single unary use or nullptr if it
+// decides not to forward.
+UnaryOp* shouldForward(Val* v) {
+  const std::vector<Expr*>& uses = v->uses();
+  // Just allow stripping out input with single use.
+  // Stripping out multi-used inputs can lead to:
+  // (1) Fragmentation of the DAG, increased segments, see test in #1301.
+  // (2) Miss detection of persistent buffers, see issue #1607.
+  if (uses.size() != 1) {
+    return nullptr;
+  }
+
+  auto* unary_use = dynamic_cast<UnaryOp*>(uses.front());
+  if (unary_use == nullptr) {
+    return nullptr;
+  }
+
+  // Don't forward an input to an output yet. Doing that would lead to an empty
+  // group that ought to work in theory but doesn't work in practice with the
+  // downstream logic. See #1813 for an example.
+  if (unary_use->out()->isFusionOutput()) {
+    return nullptr;
+  }
```

Comment:
This is the only behavior change since last snapshot. Other changes are refactors so we have only one function, shouldForward, to decide whether to forward an input.